### DISCUSSION TITLE
Reading past events with types?

### DISCUSSION THREAD

**Discussion by SleepingShell** [wevm/wagmi#1894] [Q&A]
Hi all, I'm pretty new to React and frontend development. I basically need to read all events that have ever been emitted by my contract and store them in an object (merkle tree), and then subscribe to the new events. I figure that latter will be easy to achieve with ``useContractEvent``. However, how should I best achieve reading all events once the page is loaded?

More specifically, how can I achieve type safety of events with ``getContract``? I would think this should be possible given it has access to the ABI and I generated with wagmi CLI.

Is the solution to use ABIType directly? Not sure how exactly I would do that.

**Comment by tmm** [comment]
[Stay tuned here](https://twitter.com/wagmi_sh) on Monday EST

**Comment by firnprotocol** [comment]
i am also interested in retrieving past events. i don't see a way to query for past events, either in the hooks (`"wagmi"`) or the async actions (`"wagmi/actions"`) sides of things. ethers has a way to do this, involving passing a range of blocks.

is there a way to do this that i'm missing? if not, is there a plan to add it?

thanks for the incredible work.

**Reply by zkSoju** [reply]
Any idea how to do this?

**Reply by jxom** [reply]
yea: https://viem.sh/docs/actions/public/getLogs.html

**Reply by firnprotocol** [reply]
is there an analogue in the world of `"wagmi/actions"`...?

**Reply by wildanzrrr** [reply]
is there any way to get event logs on server side?

**Comment by zoubair-hattab** [comment]
Hello guys,is there any way about that please

**Reply by firnprotocol** [reply]
try this:
```javascript
export function MyComponent() {
  const config = useConfig();
  const publicClient = usePublicClient();

  useEffect(() => {
    getBlock(config).then((block) => {
      return publicClient.getLogs({
        address: MY_ADDRESS,
        event: parseAbiItem("event MyEvent(address indexed, string)"),
        fromBlock: block.number - 999n,
        toBlock: block.number,
      });
    }).then((logs) => {
      logs.forEach((log) => {
        const topics = decodeEventLog({
          abi: MY_ABI,
          ...log,
        });
        console.log(topics);
      });
    });
  }, []);
```
which reminds me, @tmm, i noticed some unusual stuff, where swapping `event: parseAbiItem("event MyEvent(address indexed, string)"),` ----> `eventName: "MyEvent"` above caused typescript to complain. am I correct in concluding that _my approach above_, but _not_ `eventName: "MyEvent"`, is correct? what about `event: "MyEvent"`? wagmi/viem ^2.0.
